File: Java/StacksAndQueues/Check if all levels of two trees are anagrams or not.java
package StacksAndQueues;

class QueueNode<T> {
    private T data;
    private QueueNode<T> next;

    public QueueNode(T data) {
        this.data = data;
        this.next = null;
    }

    public QueueNode<T> getNext() {
        return next;
    }

    public void setNext(QueueNode<T> next) {
        this.next = next;
    }

    public T getData() {
        return data;
    }
}

class Queue<T> {
    private QueueNode<T> head;
    private QueueNode<T> tail;
    private Integer length;

    public Queue() {
        this.head = this.tail = null;
        this.length = 0;
    }

    public boolean isEmpty() {
        return this.length.equals(0);
    }

    public void push(T data) {
        QueueNode<T> node = new QueueNode<>(data);
        if (isEmpty()) {
            this.head = this.tail = node;
        }
        else {
            this.tail.setNext(node);
            this.tail = node;
        }
        this.length += 1;
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        }
        T item = this.head.getData();
        this.head = this.head.getNext();
        this.length -= 1;
        return item;
    }
}

class Node<T> {
    private T data;
    private Node<T> left;
    private Node<T> right;

    public Node(T data) {
        this.data = data;
        this.left = this.right = null;
    }

    public Node<T> getLeft() {
        return left;
    }

    public void setLeft(Node<T> left) {
        this.left = left;
    }

    public Node<T> getRight() {
        return right;
    }

    public void setRight(Node<T> right) {
        this.right = right;
    }

    public T getData() {
        return data;
    }
}

class Solution {
    public static void main(String[] args) {
        // Example 1
        Node<Integer> n1 = new Node<>(1);
        Node<Integer> n2 = new Node<>(2);
        Node<Integer> n3 = new Node<>(3);
        Node<Integer> n4 = new Node<>(4);
        Node<Integer> n5 = new Node<>(5);
        n1.setLeft(n3);
        n1.setRight(n2);
        n2.setLeft(n5);
        n2.setRight(n4);

        Node<Integer> m1 = new Node<>(1);
        Node<Integer> m2 = new Node<>(2);
        Node<Integer> m3 = new Node<>(3);
        Node<Integer> m4 = new Node<>(4);
        Node<Integer> m5 = new Node<>(5);
        m1.setLeft(m2);
        m1.setRight(m3);
        m2.setLeft(m4);
        m2.setRight(m5);

        System.out.println(areAnagrams(n1, m1));

        // Example 2
        Node<Integer> one = new Node<>(1);
        Node<Integer> two = new Node<>(2);
        Node<Integer> three = new Node<>(3);
        Node<Integer> four = new Node<>(4);
        Node<Integer> five = new Node<>(5);
        one.setLeft(two);
        two.setLeft(five);
        one.setRight(three);
        two.setRight(four);

        Node<Integer> one1 = new Node<>(1);
        Node<Integer> two1 = new Node<>(2);
        Node<Integer> three1 = new Node<>(3);
        Node<Integer> four1 = new Node<>(4);
        Node<Integer> five1 = new Node<>(5);
        one1.setLeft(two1);
        two1.setLeft(five1);
        one1.setRight(four1);
        two1.setRight(three1);

        System.out.println(areAnagrams(one, one1));
    }

    private static <T> boolean areAnagrams(Node<T> root1, Node<T> root2) {
        Queue<Node<T>> queue1 = new Queue<>();
        Queue<Node<T>> queue2 = new Queue<>();

        queue1.push(root1);
        queue2.push(root2);

        while (!queue1.isEmpty() && !queue2.isEmpty()) {
            Node<T> node1 = queue1.pop();
            Node<T> node2 = queue2.pop();

            if (!node1.getData().equals(node2.getData())) {
                return false;
            }

            if (
                    (node1.getLeft() == null && node2.getRight() != null) ||
                    (node1.getLeft() != null && node2.getRight() == null) ||
                    (node1.getRight() == null && node2.getLeft() != null) ||
                    (node1.getRight() != null && node2.getLeft() == null)
            ) {
                return false;
            }

            if (node1.getLeft() != null && node2.getRight() != null) {
                queue1.push(node1.getLeft());
                queue2.push(node2.getRight());
            }
            if (node1.getRight() != null && node1.getLeft() != null) {
                queue1.push(node1.getRight());
                queue2.push(node2.getLeft());
            }
        }
        return true;
    }
}

File: Python/Practice Set 1/Stacks and Queues/Check if all levels of two trees are anagrams or not.py
class QueueNode:
    def __init__(self, data):
        self.data = data
        self.next = None


class Queue:
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def push(self, x):
        node = QueueNode(x)
        if self.is_empty():
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1

    def pop(self):
        if self.is_empty():
            return
        item = self.head.data
        node = self.head
        self.head = self.head.next
        del node
        self.length -= 1
        return item


class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None


class Solution:
    @staticmethod
    def are_anagrams(root1: Node, root2: Node) -> bool:
        # if both the nodes are none, return true.
        if root1 is None and root2 is None:
            return True

        # if either one of them is none, while the other is not, return false.
        if root1 is None:
            return False
        if root2 is None:
            return False

        # initialize two queues and push root nodes of both the trees into them.
        queue1 = Queue()
        queue2 = Queue()
        queue1.push(root1)
        queue2.push(root2)

        # while both the queues have elements in them
        while not queue1.is_empty() and not queue2.is_empty():
            # get the nodes
            node1, node2 = queue1.pop(), queue2.pop()

            # if their data don't match, return false.
            if node1.data != node2.data:
                return False

            # if node1's left and node2's right exist, push them to the queues.
            if node1.left and node2.right:
                queue1.push(node1.left)
                queue2.push(node2.right)

            # if either one of them is none, while the other is not, return false.
            elif node1.left:
                return False
            elif node2.right:
                return False

            # same cases as above but for other pointers of nodes 1 and 2.
            if node1.right and node2.left:
                queue1.push(node1.right)
                queue2.push(node2.left)
            elif node1.right:
                return False
            elif node2.left:
                return False

        # if both the queues got empty, return true, else false.
        return queue1.is_empty() and queue2.is_empty()


# Example 1
n1, n2, n3, n4, n5 = Node(1), Node(2), Node(3), Node(4), Node(5)
n1.left = n3
n1.right = n2
n2.left = n5
n2.right = n4

m1, m2, m3, m4, m5 = Node(1), Node(2), Node(3), Node(4), Node(5)
m1.left = m2
m1.right = m3
m2.left = m4
m2.right = m5
print(Solution.are_anagrams(n1, m1))


# Example 2
one, two, three, four, five = Node(1), Node(2), Node(3), Node(4), Node(5)
one.left = two
two.left = five
one.right = three
two.right = four

one1, two1, three1, four1, five1 = Node(1), Node(2), Node(3), Node(4), Node(5)
one1.left = two1
two1.left = five1
one1.right = four1
two1.right = three1

print(Solution.are_anagrams(one, one1))
